Name: <redacted> <redacted>

Github ID: <redacted>

Github Repo Link: https://github.com/<redacted>/assignment-5-judging-flowers-<redacted>

Commit ID: 075a0ce9877193e48df18d11a4281306d2d1012c

In [29]:
# Imports and pip installations (if needed)
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn import svm
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier

# Part 1: Load the dataset

In [30]:
# Load the dataset (load remotely or locally)
data = pd.read_csv("iris.csv")

# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
print(data.head(15))
print(f"\nNo. of rows: {data.shape[0]}")
print(f"\nNo. of columns: {data.shape[0]}\n")
print(f"Classes: {data['species'].unique()}\n")
data.info()

    sepal_length  sepal_width  petal_length  petal_width species
0            5.1          3.5           1.4          0.2  setosa
1            4.9          3.0           1.4          0.2  setosa
2            4.7          3.2           1.3          0.2  setosa
3            4.6          3.1           1.5          0.2  setosa
4            5.0          3.6           1.4          0.2  setosa
5            5.4          3.9           1.7          0.4  setosa
6            4.6          3.4           1.4          0.3  setosa
7            5.0          3.4           1.5          0.2  setosa
8            4.4          2.9           1.4          0.2  setosa
9            4.9          3.1           1.5          0.1  setosa
10           5.4          3.7           1.5          0.2  setosa
11           4.8          3.4           1.6          0.2  setosa
12           4.8          3.0           1.4          0.1  setosa
13           4.3          3.0           1.1          0.1  setosa
14           5.8         

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

The dataset consists of 150 samples of iris flowers, with each sample described by five attributes: sepal_length, sepal_width, petal_length, and petal_width. These four attributes serve as the features used to predict the label, which is the species of the iris flower.

The target label species is divided into three distinct classes: 
- Iris Setosa
- Iris Versicolor
- Iris Virginica

By analyzing the relationships between the features and the label, we aim to classify each iris flower into one of these species based on its physical characteristics.

# Part 2: Split the dataset into train and test

In [31]:
# Take the dataset and split it into our features (X) and label (y)
X = data.drop('species', axis=1)
y = data['species']

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.10, random_state=45)

print(X_train)
print(X_test)
print(y_train)
print(y_test)


     sepal_length  sepal_width  petal_length  petal_width
35            5.0          3.2           1.2          0.2
145           6.7          3.0           5.2          2.3
142           5.8          2.7           5.1          1.9
40            5.0          3.5           1.3          0.3
87            6.3          2.3           4.4          1.3
..            ...          ...           ...          ...
68            6.2          2.2           4.5          1.5
95            5.7          3.0           4.2          1.2
32            5.2          4.1           1.5          0.1
124           6.7          3.3           5.7          2.1
131           7.9          3.8           6.4          2.0

[135 rows x 4 columns]
     sepal_length  sepal_width  petal_length  petal_width
0             5.1          3.5           1.4          0.2
43            5.0          3.5           1.6          0.6
129           7.2          3.0           5.8          1.6
3             4.6          3.1           1.5    

# Part 3: Logistic Regression

In [32]:
# i. Use sklearn to train a LogisticRegression model on the training set
model_LR = LogisticRegression(max_iter=200).fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
prediction = model_LR.predict(X_test.iloc[[1]]) 
print(f"Prediction: {prediction}\n")

probability = model_LR.predict_proba(X_test.iloc[[1]])
print(f"Probability for sentosa: {probability[0][0]}")
print(f"Probability for versicolor: {probability[0][1]}")
print(f"Probability for virginica: {probability[0][2]}\n")

# iii. Report on the score for Logistic regression model, what does the score measure?
score = model_LR.score(X_test, y_test) 
print(f"Score: {score}\n")

# iv. Extract the coefficents and intercepts for the boundary line(s)
coefficent = model_LR.coef_ 
print(f"Coefficent: {coefficent}\n")

intercept = model_LR.intercept_ 
print(f"Intercept: {intercept}")

Prediction: ['setosa']

Probability for sentosa: 0.9695508285975627
Probability for versicolor: 0.030448912550958003
Probability for virginica: 2.5885147919496336e-07

Score: 0.9333333333333333

Coefficent: [[-0.39865402  0.92014832 -2.45016406 -1.06107911]
 [ 0.54922877 -0.35083042 -0.13911347 -0.99339655]
 [-0.15057476 -0.5693179   2.58927752  2.05447566]]

Intercept: [  9.52861814   2.03855667 -11.56717481]


The score for the Logistic Regression Model is 0.9333, representing the accuracy of the model on the given dataset. Accuracy is calculated as the ratio of correctly predicted samples to the total samples in the test set. This means the model correctly classifies 93.33% of the instances, providing a quick measure of its performance in assigning data to the correct categories.

Equation 1: $$h(x) = 9.528 - 0.398x_1 + 0.920x_2 - 2.450x_3 - 1.061x_4$$
Equation 2: $$f(x) = 2.039 + 0.549x_1 - 0.351x_2 - 0.139x_3 - 0.993x_4$$
Equation 3: $$g(x) = -11.567 - 0.151x_1 - 0.569x_2 + 2.589x_3 + 2.054x_4$$

# Part 4: Support Vector Machine

In [33]:
# i. Use sklearn to train a Support Vector Classifier on the training set
model_svm = svm.SVC(probability=True).fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
prediction = model_svm.predict(X_test.iloc[[1]]) 
print(f"Prediction: {prediction}\n")

probability = model_svm.predict_proba(X_test.iloc[[1]])
print(f"Probability for sentosa: {probability[0][0]}")
print(f"Probability for versicolor: {probability[0][1]}")
print(f"Probability for virginica: {probability[0][2]}\n")

# iii. Report on the score for the SVM, what does the score measure?
score = model_svm.score(X_test, y_test) 
print(f"Score: {score}\n")

Prediction: ['setosa']

Probability for sentosa: 0.9494503228393472
Probability for versicolor: 0.037161797424478865
Probability for virginica: 0.013387879736174184

Score: 0.9333333333333333



The score for the SVM model indicates an accuracy of 93.33%, , meaning that the model correctly classified approximately 93.33% of the instances in the test dataset. This high score suggests that the model is performing very well in distinguishing between the classes. 

# Part 5: Neural Network

In [34]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
model_nn = MLPClassifier(random_state=1, max_iter=200).fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
prediction = model_nn.predict(X_test.iloc[[1]]) 
print(f"Prediction1: {prediction}\n")

# iii. Report on the score for the Neural Network, what does the score measure?
score = model_nn.score(X_test, y_test) 
print(f"Score1: {score}\n")

# iv: Experiment with different options for the neural network, report on your best configuration
model_nn1 = MLPClassifier(random_state=1, max_iter=10).fit(X_train, y_train)

prediction = model_nn1.predict(X_test.iloc[[1]]) 
print(f"Prediction2: {prediction}\n")

score = model_nn1.score(X_test, y_test) 
print(f"Score2 : {score}\n")


model_nn2 = MLPClassifier(random_state=1, max_iter=100).fit(X_train, y_train)

prediction = model_nn2.predict(X_test.iloc[[1]]) 
print(f"Prediction3: {prediction}\n")

score = model_nn2.score(X_test, y_test) 
print(f"Score3: {score}\n")

Prediction1: ['setosa']

Score1: 0.9333333333333333

Prediction2: ['versicolor']

Score2 : 0.26666666666666666

Prediction3: ['setosa']

Score3: 1.0



/Users/<redacted>/anaconda3/lib/python3.10/site-packages/sklearn/neural_network/_multilayer_perceptron.py:684: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/Users/<redacted>/anaconda3/lib/python3.10/site-packages/sklearn/neural_network/_multilayer_perceptron.py:684: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (10) reached and the optimization hasn't converged yet.
  warnings.warn(
/Users/<redacted>/anaconda3/lib/python3.10/site-packages/sklearn/neural_network/_multilayer_perceptron.py:684: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (100) reached and the optimization hasn't converged yet.
  warnings.warn(


Prediction 1 represents the best configuration, as Prediction 2 has a low score that may result in underfitting. In contrast, Prediction 3 achieves a perfect score, which poses a risk of overfitting during prediction.

# Part 6: K-Nearest Neighbors

In [35]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
model_knn = KNeighborsClassifier(n_neighbors=3).fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
prediction = model_knn.predict(X_test.iloc[[1]]) 
print(f"Prediction: {prediction}\n")

# iii. Report on the score for kNN, what does the score measure?
score = model_knn.score(X_test, y_test) 
print(f"Score: {score}\n")

Prediction: ['setosa']

Score: 0.9333333333333333



A K-Nearest Neighbors (KNN) score of 0.9333 indicates that the model has an accuracy of 93.33%, meaning that it correctly classified approximately 93.33% of the instances in the test dataset. This high score suggests that the KNN model is effectively distinguishing between the classes in the dataset, making reliable predictions for most of the instances.

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

All the models—Logistic Regression, SVM, MLP, and k-NN—achieved the same accuracy score of 0.93 on the dataset, but the Logistic Regression model stands out as the best choice due to its simplicity. Since all models performed equally well in terms of accuracy, the least complex model is preferred as it is computationally efficient, easier to interpret, and less prone to overfitting. It was surprising that despite the differences in model complexity, each model yielded the same performance, highlighting that simpler models can often match the effectiveness of more complex ones when the dataset is well-structured.